### PR TITLE
Add tab coloring by server group

### PR DESCRIPTION
**Pull request by MattIrv**
This PR addresses issues #327 and #63 by adding a colored line at the top of tabs when they are connected to a database, and that database is part of a server group. The color of the line is determined by the server group's color, and is 1px when the tab is inactive and 2px when the tab is active. The line does not display in high-contrast themes. It can be disabled with the setting `sql.enableTabColors`

screen shot 2017-12-20 at 1 53 33 pm

### CODE REVIEW

**Comment by kburtram** [comment]
@MattIrv I think the line might be too subtle.  Maybe we should change the entire background of the tab..or if that's too much we can think of something like 

1. Draw a line at top of tab with server group color like you have now
2. Change background color of tab to server group color but set opacity to ~20%

For reference here's how SSMS does a similar feature with user defined status bar color.

screen

File: src/sql/parts/connection/common/connectionManagement.ts [hunk 1]
```diff
@@ -220,6 +220,8 @@ export interface IConnectionManagementService {
 
 	canChangeConnectionConfig(profile: ConnectionProfile, newGroupID: string): boolean;
 
+	getTabColorForUri(uri: string): string;
```
Comment (kburtram) [inline src/sql/parts/connection/common/connectionManagement.ts]: we really need to refactor this class...it seems like the 'grab bag` for misc. (Not really related to this PR though)

File: src/sql/parts/connection/common/connectionManagementService.ts [hunk 2]
```diff
@@ -1324,4 +1334,19 @@ export class ConnectionManagementService implements IConnectionManagementService
...
+
+	public getTabColorForUri(uri: string): string {
+		if (!WorkbenchUtils.getSqlConfigValue<string>(this._workspaceConfigurationService, 'enableTabColors')) {
+			return undefined;
+		}
+		let connectionProfile = this.getConnectionProfile(uri);
+		if (!connectionProfile) {
+			return undefined;
+		}
+		let matchingGroup = this._connectionStore.getGroupFromId(connectionProfile.groupId);
```
Comment (kburtram) [inline src/sql/parts/connection/common/connectionManagementService.ts]: what do we do about connection in the root directory?  presumably they draw with the default tab color?

File: src/sql/parts/connection/common/connectionManagementService.ts [hunk 3]
```diff
@@ -148,6 +148,16 @@ export class ConnectionManagementService implements IConnectionManagementService
 
 		this.disposables.push(this._onAddConnectionProfile);
 		this.disposables.push(this._onDeleteConnectionProfile);
+
+		// Refresh editor titles when connections start/end/change to ensure tabs are colored correctly
+		let refreshEditorTitlesFunction = () => {
```
Comment (kburtram) [inline src/sql/parts/connection/common/connectionManagementService.ts]: should we have a hook to update the tab colors when the user edits the server group color, or is that a case we're not too worried about?

File: src/sql/parts/connection/common/connectionManagementService.ts (hunk 2 above)
Comment (MattIrv) [reply src/sql/parts/connection/common/connectionManagementService.ts]: Yes, those just get drawn with the default color. They'll look just like the "User Settings" tab in my screenshot

File: src/sql/parts/connection/common/connectionManagementService.ts (hunk 3 above)
Comment (MattIrv) [reply src/sql/parts/connection/common/connectionManagementService.ts]: Good idea, I'll take a look at doing that

**Comment by MattIrv** [comment]
@kburtram Here's a mockup with both of your proposals. I think drawing the background with some opacity is probably the better option between the two, but not sure how it compares to just drawing the border. What do you think?

screen shot 2017-12-20 at 2 25 15 pm

**Comment by kburtram** [comment]
@MattIrv I personally think with the opacity filter looks best of the 3.

**Comment by MattIrv** [comment]
I updated it to use the opacity filter. I also made it slightly less transparent when the tab is active, since otherwise it was difficult to tell which tab was active.

I'm still not sure the design is quite consistent with the rest of SQL Ops Studio, so I'll revisit it with @smithaBharadwaj and others who are interested after the holidays.

screen shot 2017-12-20 at 4 30 32 pm

**Comment by anthonydresser** [comment]
@MattIrv can we add a settings config for this? I'm sure there will be people who want groups but don't have the colored tabs

**Comment by MattIrv** [comment]
@anthonydresser I added the `sql.enableTabColors` setting to enable/disable it completely, but didn't add any config for whether it would fill the whole tab or not. That might be worth adding when we take another look at the design later

**Pull request merged**